core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushFilterThroughBoolOrAggregation.java

```diff
@@ -0,0 +1,293 @@
+        TupleDomain<Symbol> tupleDomain = extractionResult.getTupleDomain();
+        Expression remainingExpression = extractionResult.getRemainingExpression();
+
+        if (tupleDomain.isNone()) {
+            // Filter predicate is never satisfied. Replace filter with empty values.
+            return Result.ofPlanNode(new ValuesNode(filterNode.getId(), filterNode.getOutputSymbols(), ImmutableList.of()));
+        }
+
+        // boolOrSymbol (in remaining expressions) should only be used in Coalesce(boolOrSymbol, FALSE) expression
+        List<Expression> conjuncts = extractConjuncts(remainingExpression);
+        Map<Boolean, List<Expression>> expressions = conjuncts.stream()
+                .filter(expression -> extractUnique(expression).contains(boolOrSymbol))
+                .collect(partitioningBy(expression -> isSupportedCoalesce(expression, boolOrSymbol)));
+
+        // unsupported expressions present
+        if (!expressions.get(Boolean.FALSE).isEmpty()) {
+            return Result.empty();
+        }
+
+        Optional<Domain> boolOrDomain = Optional.ofNullable(tupleDomain.getDomains().get().get(boolOrSymbol));
+        if (boolOrDomain.isPresent() && !boolOrDomain.get().equals(Domain.singleValue(boolOrDomain.get().getType(), true))) {
+            return Result.empty();
+        }
+
+        if (expressions.get(Boolean.TRUE).isEmpty() && boolOrDomain.isEmpty()) {
+            return Result.empty();
+        }
+
+        // Push down the aggregation's symbol.
+        FilterNode source = new FilterNode(
+                context.getIdAllocator().getNextId(),
+                aggregationNode.getSource(),
+                aggregation.getArguments().getFirst());
+
+        // Remove boolOr symbol from the aggregation.
+        AggregationNode newAggregationNode = AggregationNode.builderFrom(aggregationNode)
+                .setSource(source)
+                .setAggregations(ImmutableMap.of())
+                .build();
+
+        // Add boolOr projection.
+        ProjectNode newProjectNode = new ProjectNode(
+                context.getIdAllocator().getNextId(),
+                newAggregationNode,
+                Assignments.builder()
+                        .putIdentities(source.getOutputSymbols())
+                        .put(boolOrSymbol, TRUE)
+                        .build());
+
+        // Restore identity projection if it is present in the original plan.
+        PlanNode filterSource = projectNode.map(project -> project.replaceChildren(ImmutableList.of(newProjectNode))).orElse(newProjectNode);
+
+        TupleDomain<Symbol> newTupleDomain = tupleDomain.filter((symbol, domain) -> !symbol.equals(boolOrSymbol));
+        Expression newPredicate = combineConjuncts(
+                DomainTranslator.toPredicate(newTupleDomain),
+                combineConjuncts(conjuncts.stream().filter(expression -> !extractUnique(expression).contains(boolOrSymbol)).toList()));
```

Comment:
nit: we could still replace `extractUnique(expression).contains(boolOrSymbol)` with `!equals(coalesceExpression)`.

Extracting symbols could be expensive for large expressions